src/bootstrap/bin/rustc.rs

```diff
@@ -254,7 +254,12 @@ fn main() {
         // When running miri tests, we need to generate MIR for all libraries
         if env::var("TEST_MIRI").ok().map_or(false, |val| val == "true") {
             cmd.arg("-Zalways-encode-mir");
-            cmd.arg("-Zmir-emit-validate=1");
+            // These options are preferred by miri, to be able to perform better validation,
+            // but the bootstrap compiler might not understand them.
+            if stage != "0" {
+                cmd.arg("-Zmir-emit-retag");
+                cmd.arg("-Zmir-opt-level=0");
```

Comment:
why disable optimizations in general?

Reply:
Optimizations do things like turning `&*foo` into `foo`, even if `foo` has type `&mut` -- effectively introducing a transmute. This is likely a correct optimization. However, being a correct optimization can still mean that it turns programs that used to have UB, into UB-free -- circumventing the desire of miri to find all UB.